data-pipeline/src/stats_exporter.rs

```diff
@@ -0,0 +1,600 @@
+    resource_name: String,
+    service_name: String,
+    operation_name: String,
+    span_type: String,
+    http_status_code: u32,
+    is_synthetics_request: bool,
+}
+
+/// The stats stored for each BucketKey
+#[derive(Debug, Default)]
+struct GroupedStats {
+    hits: u64,
+    errors: u64,
+    duration: u64,
+    top_level_hits: u64,
+    ok_summary: datadog_ddsketch::DDSketch,
+    error_summary: datadog_ddsketch::DDSketch,
+}
+
+impl GroupedStats {
+    fn insert(&mut self, span_stats: &SpanStats) {
+        self.duration += span_stats.duration;
+        self.hits += 1;
+
+        if span_stats.is_error {
+            self.errors += 1;
+            let _ = self.error_summary.add(span_stats.duration as f64);
+        } else {
+            let _ = self.ok_summary.add(span_stats.duration as f64);
+        }
+        if span_stats.is_top_level {
+            self.top_level_hits += 1;
+        }
+    }
+}
+
+/// Metadata required in a ClientStatsPayload
+#[derive(Debug, Default, Clone)]
+pub struct LibraryMetadata {
+    pub hostname: String,
+    pub env: String,
+    pub version: String,
+    pub lang: String,
+    pub tracer_version: String,
+    pub runtime_id: String,
+    pub service: String,
+    pub container_id: String,
+    pub git_commit_sha: String,
+    /// Should be left empty by client, except for some specific environment
+    pub tags: Vec<Tag>,
+}
+
+/// Description of a span with only data required for stats
+#[derive(Debug, Clone)]
+pub struct SpanStats {
+    pub resource_name: String,
+    pub service_name: String,
+    pub operation_name: String,
+    pub span_type: String,
+    pub http_status_code: u32,
+    pub is_synthetics_request: bool,
+    pub is_top_level: bool,
+    pub is_error: bool,
+    /// in nanoseconds
+    pub duration: u64,
+}
+
+#[derive(Debug)]
+struct StatsBucket {
+    data: HashMap<AggregationKey, GroupedStats>,
+    start: time::SystemTime,
+}
+
+impl StatsBucket {
+    fn new() -> Self {
+        Self {
+            data: HashMap::new(),
+            start: time::SystemTime::now(),
+        }
+    }
+}
+
+/// Stats exporter configuration
+#[derive(Debug, Default)]
+pub struct Configuration {
+    /// time range of each bucket
+    pub buckets_duration: time::Duration,
+    /// optional timeout for sending stats
+    pub request_timeout: Option<time::Duration>,
+    /// endpoint used to send stats to the agent
+    pub endpoint: Endpoint,
+}
+
+/// An exporter aggregating stats from traces and sending them to the agent
+///
+/// Currently we only keep one time bucket starting at the time of the exporter creation and
+/// resetting to current time on flush. All `SpanStats` sent between flushesare added to the same
+/// bucket.
+/// This raises two issues:
+/// - We expect SpanStats to be submitted right after the span ended (since the aggregation is done
+///   on endTime)
+/// - We expect the tracer to call send when we reach start_time + bucket_duration to make sure the
```

Comment:
Is this a reasonable / safe assumption to make for users of this library? Is there any reason we avoided doing this flushing on the library side? complexity with FFI / wanting to avoid buffering too many stats buckets?